xla/service/gpu/runtime/copy_thunk.cc

```diff
@@ -125,6 +125,11 @@ absl::Status DeviceToHostCopyThunk::ExecuteOnStream(
   TF_ASSIGN_OR_RETURN(
       se::Stream * stream,
       GetStreamForExecution(Thunk::execution_stream_id(), params));
+  // Wait for the source to be ready on device unless it's an input parameter
+  if (instr_->operand(0)->opcode() != HloOpcode::kParameter) {
```

Comment:
It is unsafe to access `instr_` (it should be removed from thunk), because nothing guarantees that HloModule will be alive at run time, it's accidental that it's usually available.

Also instead of `stream->WaitFor` in a thunk, it should a be a `WaitFor` thunk in a sequence that enforces execution order and adds necessery stream syncs. It's a bug in IrEmitterUnnested that CopyThunk was emitted without a stream wait thunk.

Reply:
Basically `copy-start` and `copy-done` lowering are responsible for adding stream wait thunks to guarantee that nested copy is correctly syncrhonized

Reply:
Thanks @ezhulenev for pointing out the difference between `stream->WaitFor` and a `WaitFor` thunk. I'll rewrite the code to add a `WaitFor` thunk.